src/relay/op/nn/nn.cc

```diff
@@ -15,6 +15,112 @@
 namespace tvm {
 namespace relay {
 
+TVM_REGISTER_NODE_TYPE(DenseAttrs);
+
+
+bool DenseRel(const Array<Type>& types,
+              int num_inputs,
+              const Attrs& attrs,
+              const TypeReporter& reporter) {
+  CHECK_EQ(types.size(), 3);
+  const auto* data = types[0].as<TensorTypeNode>();
+  const auto* weight = types[1].as<TensorTypeNode>();
+  if (data == nullptr) return false;
+
+  const DenseAttrs* param = attrs.as<DenseAttrs>();
+  CHECK(param != nullptr);
+
+  CHECK(static_cast<int>(data->shape.size()) != 0);
+
+  Array<tvm::Expr> oshape = data->shape;
+  if (param->units.defined()) {
+    Array<tvm::Expr> dshape = data->shape;
+
+    // validate the weight shape is proper if defined
+    if (weight != nullptr) {
+      CHECK(reporter->AssertEQ(weight->shape[0], dshape[dshape.size() - 1]))
+        << "Dense: shape of weight is inconsistent with input data.";
+      CHECK(reporter->AssertEQ(weight->shape[1], param->units))
```

Comment:
Don't need to check weight !=nullptr, The Assign will run the check internally. Directly goto the assign logic is fine